Create superpmi-asmdiffs pipeline

BruceForstall opened the pull request:
Create a new `runtime-coreclr superpmi-asmdiffs` pipeline that runs SuperPMI asmdiffs for every change in the JIT directory.

The diffs are run on two platforms: Windows x64 and Windows x86. Linux, and Arm64 and Arm32, asm diffs are done using cross-compilers, as follows:

| Platform | Asm diffs |
| -- | -- |
| Windows x64 | win-x64, win-arm64, linux-x64, linux-arm64 |
| Windows x86 | win-x86, linux-arm |

The resulting summary .md files are uploaded into the pipeline artifacts, one .md file per platform (so, one for the Windows x64 runs and one for the Windows x86 runs). The results are also displayed in "Extensions" page of the AzDO pipeline.

It looks like the runs take about 50 minutes to complete (assuming not much waiting for machines).

The asm diffs pipeline is similar to the "superpmi-replay" pipeline, except:
1. It determines what an appropriate baseline JIT would be based on the PR commit and how it merges with the `main` branch. Given this, it downloads the matching baseline JITs from the JIT rolling build artifacts in Azure Storage.
2. It clones the `jitutils` repo and builds the `jit-analyze` tool, needed to generate the summary .md file.
3. It downloads and adds to the Helix machine payload a "portable" `git` installation, as `git diff` is used by `jit-analyze` for analyzing the generated .dasm files of the diff.
4. It collects all the various summary.md files into one per platform on which the runs are done, and publishes that to the artifacts and the `Extensions` page.
5. It only does one replay (asmdiffs) run, not one for each of a set of multiple stress modes.

As part of this implementation,
a. The `azdo_pipelines_util.py` was renamed to `jitutil.py`, and a lot of utility functions from superpmi.py were moved over to it. This was mostly to share the code for downloading and uncompressing .zip files. (There might be some slight changes to the output from the superpmi.py download commands that I'll have to look into.) However, I also moved a bunch of simple, more general helpers, for possible future sharing.
b. `jitrollingbuild.py download` can now take no arguments and download a baseline JIT (from the JIT rolling build Azure Storage location), for the current enlistment, to the default location. Previously, it required a specific git_hash and target directory. There is similar logic in superpmi.py, but not quite the same.
c. The `superpmi.py --no_progress` option was made global, and applied in a few more places.

Fixes https://github.com/dotnet/runtime/issues/59445

BruceForstall committed:
Create superpmi-asmdiffs pipeline

BruceForstall committed:
Set -error_limit

BruceForstall committed:
Move baseline download to setup script

BruceForstall committed:
Use `origin/main` for baseline, not `main`

BruceForstall committed:
Fix directory creation

BruceForstall committed:
Make `--no_progress` a global superpmi.py switch

`asmdiffs` can download coredistools, for example, so don't
output download progress if the user wants to suppress it.

Also, move the `download` log file to the "uploads" area so it gets uploaded
to the AzDO files.

Temporarily, only do diffs with benchmarks, so speed things up

BruceForstall committed:
Add support for running jit-analyze on the diffs

BruceForstall committed:
Add summarizing of asm diffs

BruceForstall committed:
Re-enable diffing all MCH files

BruceForstall committed:
Rename azdo_pipelines_util.py -> jitutil.py

BruceForstall committed:
Move many functions from superpmi.py to jitutil.py

Add downloading of portable git for use on Helix for jit-analyze.
It is added to the correlation payload that is copied to Helix machines
(currently, we only use Windows Helix machines).

BruceForstall committed:
Run summarize before uploading log files

Overall summary.md file should be uploaded as well.

BruceForstall committed:
Fix summarize to walk the tree

BruceForstall commented:
Sample run AzDO results with forced diffs (loop cloning disabled for diff compiler), with summary in the "Extensions" tab:

https://dev.azure.com/dnceng/public/_build/results?buildId=1457493&view=ms.vss-build-web.run-extensions-tab

and the published artifacts contain the "overall summary" Markdown files in the `SuperPMI_Logs_<arch>_checked` folders:

https://dev.azure.com/dnceng/public/_build/results?buildId=1457493&view=artifacts&pathAsName=false&type=publishedArtifacts

BruceForstall committed:
Remove change to force asm diffs

Add a few more comments

BruceForstall commented:
@dotnet/jit-contrib This is ready for review. PTAL.

BruceForstall commented:
Some ideas for improvements:

- If there are no diffs, make that explicit, don't just show empty "overall_....md" files in the Extensions page
- Create summary.md files for other metrics, especially PerfScore. Maybe these should be summarized as one liners in the Extensions page, but not fully expanded, to prevent that page from getting too cluttered. Make them available for download.
- Improve the headers in the overall summary output, e.g., add per-platform headers.
- Q: can there be multiple levels of "disclosure" / tree in the output, so we can "expand" to view just one platform at a time, to prevent clutter?
- Make sure that JIT assertions, if any, are shown or at least reported very prominently in the Extensions output.
- Figure out why the recent runs aren't getting good data when looking for a JIT baseline (e.g., https://dev.azure.com/dnceng/public/_build/results?buildId=1457968&view=logs&j=011e1ec8-6569-5e69-4f06-baf193d1351e&t=bf6cf4cf-6432-59cf-d384-6b3bcf32ede2)

AndyAyersMS commented:
The extensions thing is pretty cool; I had no idea that existed.

I find the current format requires a lot of scrolling to fully comprehend -- there is a lot of data and the interesting bits can be buried. It would be nice to have everything in one table and perhaps sorted by largest impact or some such. Doesn't work as well for detail/expand collapse but perhaps we can get to this via internal links or something?

Also (assuming interesting/ unexpected diffs appear) is it obvious how to repro exactly what the CI did? 

Maybe echo out the repro commands somewhere (including exact base jit version used / exact collections used if we ever start versioning beyond the jit guid)

BruceForstall commented:
> I find the current format requires a lot of scrolling to fully comprehend

The example I gave ([here](https://dev.azure.com/dnceng/public/_build/results?buildId=1457493&view=ms.vss-build-web.run-extensions-tab)) has a lot of diffs, because I disabled loop cloning to force diffs. However, we are generating diffs for 6 platforms. And we propose to also generate PerfScore diff results.

Do you have any suggestion on how to summarize this better? Kunal suggested something like "Impact" table [here](https://github.com/dotnet/runtime/pull/60787#issue-1033969670).

Printing out repro commands makes sense. Can probably specify the precise baseline git hash used, but assume the diff is the current tree, and then just specify the correct `-arch`, `-target_arch`, `-target_os` options.

Note, however, I don't want to gate this change on creating the "perfect" output.

AndyAyersMS commented:
> Do you have any suggestion on how to summarize this better? Kunal suggested something like "Impact" table here.

Yes, something very much like that would be great.

> Note, however, I don't want to gate this change on creating the "perfect" output.

Agreed -- no need to hold this up based on my feedback; it is quite useful even as-is (eg trigger on community PRs that we think should be no diff). As we get mileage on it we will figure out enhancements to make it more useful.

BruceForstall commented:
> eg trigger on community PRs that we think should be no diff

Currently it's set to trigger on _every_ PR that touches the JIT directory. I think we should keep that level of checking.

kunalspathak commented:
One more optimization we spoke about it to skip creating `.dasm` files altogether and just return the metrics numbers but it is currently blocked by https://github.com/dotnet/runtime/issues/52877.

BruceForstall committed:
Merge branch 'main' of https://github.com/dotnet/runtime into runtime-asmdiffs

kunalspathak commented on src/coreclr/scripts/superpmi_asmdiffs.py:
```diff
@@ -0,0 +1,185 @@
+        "-log_level", "debug",
+        "-log_file", log_file])
+
+    # If there are asm diffs, and jit-analyze ran, we'll get a diff_summary.md file in the spmi_location directory.
```
Isn't that called `summary.md`?
https://github.com/dotnet/runtime/blob/dfb8c90b86ec67ffc1192a1561ce11e281ee9619/src/coreclr/scripts/superpmi.py#L2015-L2018

BruceForstall replied:
Each MCH file creates a summary.md; superpmi.py appends these all together into a `diff_summary.md` file (so, one `diff_summary.md` file for each platform/arch run).

kunalspathak commented on src/coreclr/scripts/jitutil.py:
```diff
@@ -0,0 +1,666 @@
+#!/usr/bin/env python3
```
Can you point to the code that was modified that needs reviewed? If we just moved code to this file, then I will skip reviewing it.

BruceForstall replied:
It's mostly moved code. Some changes:
1. copy_directory now has `verbose_copy` and `verbose_skip` options, to be more granular with the verbosity.
2. some of the functions like `download_one_url` take an `is_azure_storage` argument stating whether the download is coming from Azure blob storage or not (in which case we assume it's a URL or possibly a Windows UNC path).

The biggest thing here is I moved over the download_files function from superpmi.py so I could use its functionality to download a ZIP file, uncompress it, and copy all its contents to a target.

kunalspathak commented on src/coreclr/scripts/jitrollingbuild.py:
```diff
@@ -172,6 +182,111 @@ def determine_jit_name(coreclr_args):
+    with ChangeDir(coreclr_args.runtime_repo_location):
+        command = [ "git", "rev-parse", "HEAD" ]
+        print("Invoking: {}".format(" ".join(command)))
+        proc = subprocess.Popen(command, stdout=subprocess.PIPE)
```
we should try to use `run_command` instead. It has pretty much most of the functionality of getting back stdout, return code, etc.

BruceForstall replied:
The trick with using `run_command` is it has a different output model. We could introduce a `verbose_output` option. I propose to defer changing this.

BruceForstall committed:
Make sure `git fetch` and jitrollingbuild.py are run from the source directory

kunalspathak commented on src/coreclr/scripts/jitrollingbuild.py:
```diff
@@ -172,6 +182,111 @@ def determine_jit_name(coreclr_args):
+        command = [ "git", "log", "--pretty=format:%H", base_git_hash, "-20", "--", "src/coreclr/jit/*" ]
+        print("Invoking: {}".format(" ".join(command)))
+        proc = subprocess.Popen(command, stdout=subprocess.PIPE)
+        stdout_change_list, _ = proc.communicate()
```
Worth printing the output of `stdout_change_list` for ease debugging.

BruceForstall replied:
We do print each hash that we look at in the returned output. Maybe that's sufficient in the normal case?

kunalspathak commented on src/coreclr/scripts/superpmi_asmdiffs_summarize.py:
```diff
@@ -0,0 +1,94 @@
+
+    # Consolidate all superpmi_diff_summary_*.md in overall_diff_summary_<platform>_<architecture>.md
+    # (Don't name it "superpmi_xxx.md" or we might consolidate it into itself.)
+    final_md_path = os.path.join(diff_summary_dir, "overall_diff_summary_{}_{}.md".format(platform, arch))
```
We should also print the baseline git-hash here, probably using full url, to make sure that developer knows which baseline was used for comparison.

BruceForstall replied:
We don't actually have that information here. We could presumably pass it back from the Helix machine, but I think that should be left for future improvement. The baseline can be found in the AzDO log files if actually needed. I presume most people won't need to know the exact baseline hash to repro diffs.

kunalspathak reviewed (commented):
Thanks for taking this work. Added few suggestions/comments.

BruceForstall commented on eng/pipelines/coreclr/templates/run-superpmi-replay-job.yml:
```diff
@@ -65,9 +65,16 @@ jobs:
-    - script: |
-        mkdir -p $(SpmiCollectionLocation)
-      displayName: Create directory for SPMI collection
+    - ${{ if ne(parameters.osGroup, 'windows') }}:
```
`mkdir -p foo` on windows creates a `-p` directory and a `foo` directory.

We could remove the `osGroup` checking, but I figured we might as well be consistent everywhere and check it just in case we expand to support running on non-Windows OS? I was trying to make the various YML files consistent.

kunalspathak replied:
My take is we can add the various `osGroup` checks in future when we make this compatible for non-Windows OS. For now, it just adds maintenance overhead so I would just remove everything for non-Windows.

kunalspathak replied:
Another reason I don't think we should include unnecessary things in these yml is mostly we end up copying them to create new pipeline and the irrelevant things get duplicated in new files as well.

BruceForstall commented on eng/pipelines/coreclr/templates/run-superpmi-asmdiffs-job.yml:
```diff
@@ -0,0 +1,133 @@
+    steps:
+    - ${{ parameters.steps }}
+
+    - ${{ if ne(parameters.osGroup, 'windows') }}:
```
Yes, but I thought I'd leave both osGroup/osSubGroup and archType in case we expand to run on non-Windows

BruceForstall commented on src/coreclr/scripts/superpmi_asmdiffs_summarize.py:
```diff
@@ -0,0 +1,94 @@
+        for dirpath, _, files in os.walk(diff_summary_dir):
+            for file_name in files:
+                if file_name.startswith("superpmi_") and file_name.endswith(".md"):
+                    full_file_path = os.path.join(dirpath, file_name)
```
I've added fancier output (will push PR with an example)

BruceForstall commented on src/coreclr/scripts/superpmi_asmdiffs_summarize.py:
```diff
@@ -0,0 +1,94 @@
+
+    diff_summary_dir = coreclr_args.diff_summary_dir
+    arch = coreclr_args.arch
+    platform = coreclr_args.platform
```
Currently, yes, but I thought I'd keep it for generality.

BruceForstall commented on src/coreclr/scripts/superpmi_asmdiffs_setup.py:
```diff
@@ -0,0 +1,254 @@
+        os.makedirs(base_jit_directory)
+
+    print("Fetching history of `main` branch so we can find the baseline JIT")
+    run_command(["git", "fetch", "origin", "main"], _exit_on_fail=True)
```
When I forced setting the directory before calling `git fetch`, it seemed to fix my problem.

As for fetching a subset: the clone of the repro (on my machine, using the commands this pipeline uses) takes 68 seconds and takes 1.2GB. Doing the `git fetch main` takes 2 seconds and consumes an additional 3.6MB. As such, I don't think we need to restrict what `git fetch` fetches.

BruceForstall commented on src/coreclr/scripts/superpmi-asmdiffs.proj:
```diff
@@ -0,0 +1,73 @@
+    <!-- Workaround until https://github.com/dotnet/arcade/pull/6179 is not available -->
+    <HelixResultsDestinationDir>$(BUILD_SOURCESDIRECTORY)\artifacts\helixresults</HelixResultsDestinationDir>
+    <WorkItemCommand>$(Python) $(ProductDirectory)\superpmi_asmdiffs.py -base_jit_directory $(ProductDirectory)\base -diff_jit_directory $(ProductDirectory)\diff -log_directory $(SuperpmiLogsLocation)</WorkItemCommand>
+    <WorkItemTimeout>3:15</WorkItemTimeout>
```
I've seen it take as long as 1:40, mostly due to waiting for machines. Mostly, it seems to be around 1:10. I'll make the timeout 1:45 and we can see how that goes.

kunalspathak replied:
The `WorkItemTimeout` is the actual timeout of the work that helix performs. So once `superpmi asmdiff` starts for all the collections, it doesn't take more than 45 minutes to complete. So probably 60 minutes is reasonable here.

BruceForstall commented on src/coreclr/scripts/superpmi_asmdiffs.py:
```diff
@@ -0,0 +1,185 @@
+        python_path,
+        os.path.join(script_dir, "superpmi.py"),
+        "asmdiffs",
+        "--no_progress",
```
Mostly because asmdiffs will download coredistools. Everything else (baseline JIT, MCH files) has already been downloaded.

BruceForstall commented on eng/pipelines/coreclr/superpmi-asmdiffs.yml:
```diff
@@ -0,0 +1,30 @@
+  paths:
+    include:
+    - src/coreclr/jit/*
+    - src/coreclr/inc/jiteeversionguid.h
```
Good point; I'll remove that.

BruceForstall committed:
Feedback

1. Drop timeout to 1:45
2. Create better output summary report
3. Don't run on JIT-EE interface changes (if a JIT change also includes
a JIT-EE interface change, it will still run, and it will be expected
to fail. I'm not too worried about this case.)

BruceForstall commented:
@kunalspathak @AndyAyersMS Here's a sample of what the asmdiffs reports look like now:

https://github.com/dotnet/runtime/pull/61383/files?short_path=b3cf220#diff-b3cf2208bc362d31e2bcc50677015e80b5a448261fd9f9f562b1c1cb3e5b3c13

BruceForstall committed:
Only be verbose on copies

This affects `superpmi.py download` output

BruceForstall commented:
@kunalspathak, others: I updated this change for the feedback.

BruceForstall commented:
Here's the Extensions page result we now generate if there are no diffs:

https://dev.azure.com/dnceng/public/_build/results?buildId=1462505&view=ms.vss-build-web.run-extensions-tab

This run completed successfully except the aspnet collection replay has MISSING data, and that is reported as an error code.

BruceForstall committed:
Assume Windows; don't use osGroup or `-platform` option to summarize script

There will be more changes required if we ever run on non-Windows platforms,
so don't keep these partial measures.

BruceForstall committed:
Reduce work item timeout to 1:00

kunalspathak reviewed (approved):
LGTM. Thanks for doing this, can't wait to trigger it on one of my PRs.

kunalspathak commented on eng/pipelines/coreclr/templates/run-superpmi-asmdiffs-job.yml:
```diff
@@ -0,0 +1,127 @@
+      inputs:
+        sourceFolder: '$(HelixResultLocation)'
+        contents: '**/superpmi_*.md'
+        targetFolder: '$(SpmiLogsLocation)'
```
I just realized that we should have a different name for folder that contains `.md` files. Something like `SuperPMI_Asmdiffs_$(archType)_$(buildConfig)`.

BruceForstall replied:
Sure, I can do that.

kunalspathak commented:
I don't see the logic to upload other metric `summary.md`, e.g. `PerfScore`. Are you planning to do that in a follow-up PR?

AndyAyersMS commented:
> Thanks for doing this, can't wait to trigger it on one of my PRs.

Agree -- thanks for working on this. 

👍

BruceForstall commented:
> I don't see the logic to upload other metric `summary.md`, e.g. `PerfScore`. Are you planning to do that in a follow-up PR?

True, I haven't done that.

I can pass `-metrics CodeSize -metrics PerfScore` to superpmi.py to add PerfScore to the summary.md. Then, PerfScore and CodeSize will be mixed together in the MD files. IMO, that's ok, even if it potentially creates additional "clutter". However, it disables the recent work to use the "actual" code size instead of just the code size for functions with diffs, which is unfortunate. See https://github.com/dotnet/runtime/pull/61254#issuecomment-965645177.

Generating the PerfScore metrics separately would require iterating over all the respective MCH directories, invoking jit-analyze specifically for each base/diff pair, and then collecting a summary_PerfScore.md file, which would be separately summarized and uploaded. This is more work than I want to do.

So I think for now I'm not going to add additional metrics to the summary.

BruceForstall committed:
Move asmdiffs .md files to separate log location

Also, reduce timeout values

kunalspathak commented:
> However, it disables the recent work to use the "actual" code size instead of just the code size for functions with diffs, which is unfortunate. 

What do you mean? It will print the `PerfScore` diff at least, right? It won't be accurate with actual PerfScore of everything combined?

BruceForstall commented:
> What do you mean? It will print the `PerfScore` diff at least, right? It won't be accurate with actual PerfScore of everything combined?

The implementation of the accurate CodeSize doesn't kick in when any metrics are specifically requested: https://github.com/dotnet/runtime/blob/1d352fc8164a43e787cb17782ae26d17e5e28b13/src/coreclr/scripts/superpmi.py#L2030-L2033

There is no "accurate" total PerfScore number currently, just CodeSize.

kunalspathak commented:
Can we do something like this in `superpmi.py` itself? I am mostly interested in PerfScore and trying to see easy way to get it in along with this PR.

```py
      command = [ jit_analyze_path, "--md", md_summary_file, "-r", "--base", base_asm_location, "--diff", diff_asm_location ]

      if base_bytes is not None and diff_bytes is not None:
          m_command = command + [ "--override-total-base-metric", str(base_bytes), "--override-total-diff-metric", str(diff_bytes) ]

      run_and_log(m_command, logging.INFO)

    # run again for non-code size metrics.
     non_code_size_metrics = [m for m in self.coreclr_args.metrics if m != 'CodeSize']
      if len(non_code_size_metrics ) > 0:
          m_command = command + [ "--metrics", ",".join(non_code_size_metrics) ]
          run_and_log(m_command, logging.INFO)
```

kunalspathak commented:
By the way, I didn't realize until now that we can pass `-metric` to superpmi.py script.

BruceForstall commented:
Something like that would work, but we'd then have two .md files, or have to force the second jit-analyze to append to the first one.

I think a better solution is to pass "override" data on a per-metric basis. That is, teach jit-analyze to understand `--override-total-base-metric <metric>,<value>` instead of just `--override-total-base-metric <value>` (and support multiple overrides, one per metric). Then, superpmi.py could invoke jit-analyze exactly once. And for now, just pass the CodeSize override arguments.

A not-mutually-exclusive alternative is to teach superpmi.py to (optionally?) create one summary.md file for each metric, e.g., `summary_<metric>.md`, by either invoking jit-analyze multiple times, or teaching jit-analyze to split the results (the latter would presumably be much faster).

kunalspathak commented:
> or teaching jit-analyze to split the results (the latter would presumably be much faster).

I agree. We can just ask `jit-analyze` to dump the metrics to `<md_filename_from_args>.metrics_name.md`.

But then, we need to still change the logic around passing `--override-total-base-metric ` to jit-analyze, right?

BruceForstall commented:
> But then, we need to still change the logic around passing `--override-total-base-metric ` to jit-analyze, right?

Yes, or generalize it as I suggest.

kunalspathak commented:
> Yes, or generalize it as I suggest.

Ok, whatever is easy and clean is fine then.
